src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,577 @@
+                Objects.requireNonNull(initializer),
+                Objects.requireNonNull(integrator),
+                Objects.requireNonNull(combiner),
+                Objects.requireNonNull(finisher)
+        );
+    }
+
+    /**
+     * A Downstream object is the next stage in a pipeline of operations,
+     * to which elements can be sent.
+     * @param <T> the type of elements this downstream accepts
+     * @since 22
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Downstream<T> {
+
+        /**
+         * Pushes, if possible, the provided element to the next stage in the
+         * pipeline.
+         *
+         * <p>If this method returns {@code false} then the next stage does
+         * not accept any more elements.
+         *
+         * @param element the element to send
+         * @return {@code true} if more elements can be sent,
+         *         and {@code false} if not.
+         */
+        boolean push(T element);
+
+        /**
+         * Allows for checking whether the next stage is known to not want
+         * any more elements sent to it.
+         *
+         * @apiNote This is best-effort only, once this returns true it should
+         *          never return false again for the same instance.
+         *
+         * <p>By default this method returns {@code false}.
+         *
+         * @return {@code true} if this Downstream is known not to want any
+         *         more elements sent to it, {@code false} if otherwise
+         */
+        default boolean isRejecting() { return false; }
+    }
+
+    /**
+     * An Integrator receives elements and processes them,
+     * optionally using the supplied state, and optionally sends incremental
+     * results downstream.
+     *
+     * @param <A> the type of state used by this integrator
+     * @param <T> the type of elements this integrator consumes
+     * @param <R> the type of results this integrator can produce
+     * @since 22
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Integrator<A, T, R> {
+        /**
+         * Performs an action given: the current state, the next element, and
+         * a downstream object; potentially inspecting and/or updating
+         * the state, optionally sending any number of elements downstream
+         * -- and then returns whether more elements are to be consumed or not.
+         *
+         * @param state The state to integrate into
+         * @param element The element to integrate
+         * @param downstream The downstream object of this integration,
+         *                   returns false if doesn't want any more elements
+         * @return {@code true} if subsequent integration is desired,
+         *         {@code false} if not
+         */
+        boolean integrate(A state, T element, Downstream<? super R> downstream);
```

Comment:
Integrator is silent on nulls and whether NPE is thrown.

Reply:
No sure what it should say regarding nulls. The `initializer` might return `null` in which case `state` would be `null`.  `element` can be `null` if the stream contains `null`s. `Downstream` should never be null.

Should the Javadoc say something about how RuntimeExceptions thrown must be interpreted by the caller? 🤔